### PR TITLE
Removes misleading "No such file or directory" syntax errors from output files

### LINKED ISSUE #1252: [NCO Bug] Remove misleading "No such file or directory" syntax errors from output files
Bugzilla [#1369](http://www2.spa.ncep.noaa.gov/bugzilla/show_bug.cgi?id=1369)
Bugzilla [#1016](http://www2.spa.ncep.noaa.gov/bugzilla/show_bug.cgi?id=1016)

Details from bugzilla:

From NCO SPA:
```
In the output logs for gfs/gdas/enkfgdas there are hundres of instances of "No such file or directory" messages 
printed out across many different tasks, which makes it more difficult to troubleshoot production failures.
Please work towards removing all related syntax errors from the output logs

A list of tasks from the 12z cycle found with these messages:
enkfgdas_diag_12
enkfgdas_fcst_??_12
gdas_atmos_analysis_diag_12
gdas_atmos_tropcy_qc_reloc_12
gdas_atmos_verfozn_12
gdas_atmos_verfrad_12
gdas_forecast_12
gfs_atmos_gempak_meta_12
gfs_atmos_tropcy_qc_reloc_12
gfs_atmos_wafs_grib2_0p25_12
gfs_atmos_wafs_grib2_12
gfs_mos_ext_grd_prdgen_12
gfs_mos_ext_grd_prep_12
gfs_mos_ext_stn_fcst_12
gfs_mos_ext_stn_prdgen_12
gfs_mos_ext_stn_prep_12
gfs_mos_grd_fcst_12
gfs_mos_grd_prdgen_12
gfs_mos_stn_fcst_12
gfs_mos_stn_prdgen_12
gfs_mos_wx_ext_prdgen_12
gfs_mos_wx_prdgen_12
gfs_wave_prdgen_gridded_12
```

Additional jobs from the GEFS v12 package:
```
gefs_atmos_prdgen_gfs_00.o37433087:rm: cannot remove 'exlist': No such file or directory
gefs_forecast_0_16d_c00_00.o37433286:ls: cannot access /gpfs/dell4/nco/ops/nwges/para/gefs.20200916/00/init/c00/RESTART/coupler: No such file or directory
gefs_forecast_0_16d_c00_00.o37433286:ls: cannot access /gpfs/dell4/nco/ops/com/gefs/para/gefs.20200916/00/wave/rundata/rmp_src_to_dst_conserv_*: No such file
or directory
gefs_wave_prep_c00_00.o37432963:mv: cannot stat 'ice.out': No such file or directory
```

Comment from @RussTreadon-NOAA:
```
A check of the gdas_atmos_analysis_diag and enkfgdas_diag "No such file" messages shows that the messages 
originate from the same script, exglobal_diag.sh.  The following loop is responsible for generating the 
"No such file" messages

   # Restrict diagnostic files containing rstprod data
   rlist="conv_gps conv_ps conv_pw conv_q conv_sst conv_t conv_uv saphir"
   for rtype in $rlist; do
       ${CHGRP_CMD} *${rtype}*
   done

A check on file existence can be added to ensure ${CHGRP_CMD} is only executed when ${rtype} exists.

For GFS v16.3 exgdas_diag.sh resides in NOAA-EMC/GSI branch release/gfsda.v16.3.0.  Note that NOAA-EMC/GSI 
develop no longer maintains DA jobs or scripts.  GFS DA jobs and scripts are now maintained in the 
global-workflow repository.  Therefore, add Rahul Mahajan to the cc list for this bugzilla.
```

Comment from Steven Earle:
```
gfs_mos tasks are excluded from this... but please do a thorough review of all gfs/gdas/enkf tasks for 
version 17... we need to work toward remove all false/misleading errors from our output.
```

### PR DESCRIPTION
**Pull request by HenryRWinterbottom**
# Description

This PR addresses issue #1252. The following is accomplished:

- Prior to removing files, the existence of a file is checked prior to attempting to remove; this is performed as noted [here](https://github.com/NOAA-EMC/global-workflow/issues/1252#issue-1538627369); this PR only addresses the the `chgrp` issue.

  References #1252

# Type of change

- Maintenance (code refactor, clean-up, new CI test, etc.)

# Change characteristics
- Is this a breaking change (a change in existing functionality)? NO
- Does this change require a documentation update? NO

# How has this been tested?

The updates have been tested for the following CI/CD tests on RDHPCS Hera:

- `C48_ATM.yaml`
- `C48_S2SW.yaml`
- `C96_atm3DVar_extended.yaml`

When grepping through the logs, there remain `No such file or directory` messages. However, they are produced by ancillary packages utilized within global-workflow -- namely the observation prep job(s) and the tracker and genesis packages. Further, as noted by @WalterKolczynski-NOAA, the `setpdy.sh` produces similar messages but remains beyond the scope of this PR.

# Checklist
- [x] Any dependent changes have been merged and published
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] New and existing tests pass with my changes
- [x] I have made corresponding changes to the documentation if necessary

### CODE REVIEW

File: scripts/exglobal_diag.sh [hunk 1]
```diff
@@ -233,8 +233,12 @@ EOFdiag
...
    for rtype in $rlist; do
       set +e
-      ${CHGRP_CMD} *${rtype}*
-      ${STRICT_ON:-set -e}
+      for rfile in *"${rtype}"*; do 
+	 if [[ -s "${rfile}" ]]; then
+            ${CHGRP_CMD} "${rfile}"
+	    ${STRICT_ON:-set -e}
+         fi
+      done
```
Comment (DavidHuber-NOAA) [inline scripts/exglobal_diag.sh]: Please use spaces instead of tabs for white space.

```suggestion
      for rfile in *"${rtype}"*; do 
         if [[ -s "${rfile}" ]]; then
            ${CHGRP_CMD} "${rfile}"
            ${STRICT_ON:-set -e}
         fi
      done

**Review by DavidHuber-NOAA — CHANGES_REQUESTED** [review]

**Comment by DavidHuber-NOAA** [comment]
@HenryWinterbottom-NOAA Could you update the description to specify that this is only addressing the `chgrp` issue?

I think it may make sense to keep #1252 open and/or open other issues (here and/or in obsproc and tc_tracker) to finish the work outlined in the bugzilla.  If these are already open, could you link them to #1252?

**Comment by HenryRWinterbottom** [comment]
> @HenryWinterbottom-NOAA Could you update the description to specify that this is only addressing the `chgrp` issue?
> 
> I think it may make sense to keep #1252 open and/or open other issues (here and/or in obsproc and tc_tracker) to finish the work outlined in the bugzilla. If these are already open, could you link them to #1252?

The description is updated.

My understanding, see the notes above, is that the ancillary package updates are beyond the scope of this PR.

**Comment by DavidHuber-NOAA** [comment]
@HenryWinterbottom-NOAA Agreed.  Those comments are only in regards to issue #1252, not this PR.

**Review by DavidHuber-NOAA — APPROVED** [review]
Looks good.

**Comment by HenryRWinterbottom** [comment]
> @HenryWinterbottom-NOAA Agreed. Those comments are only in regards to issue #1252, not this PR.

@DavidHuber-NOAA Sorry, I misunderstood. Thank you for clarifying.

File: scripts/exglobal_diag.sh [hunk 2]
```diff
@@ -233,8 +233,12 @@ EOFdiag
    rlist="conv_gps conv_ps conv_pw conv_q conv_sst conv_t conv_uv saphir"
    for rtype in $rlist; do
       set +e
-      ${CHGRP_CMD} *${rtype}*
-      ${STRICT_ON:-set -e}
+      for rfile in *"${rtype}"*; do 
+         if [[ -s "${rfile}" ]]; then
+            ${CHGRP_CMD} "${rfile}"
+            ${STRICT_ON:-set -e}
```
Comment (aerorahul) [inline scripts/exglobal_diag.sh]: Shouldn't this be out of the `for` loop.
If there are no files that match the `*"${rtype}"` or if the `if` condition is not satisfied, `set +e` will prevail at the end of the for loop.  `set -e` should be set there based on how it is in `develop`.

Comment (WalterKolczynski-NOAA) [reply scripts/exglobal_diag.sh]: Now that we are checking for the file's existence, I'm not sure `set +e` is needed anymore at all.

Comment (HenryRWinterbottom) [reply scripts/exglobal_diag.sh]: Please confirm that my update makes the correction(s) that both of you are suggesting.

File: scripts/exglobal_diag.sh [hunk 3]
```diff
@@ -232,9 +232,12 @@ EOFdiag
...
    rlist="conv_gps conv_ps conv_pw conv_q conv_sst conv_t conv_uv saphir"
    for rtype in $rlist; do
-      set +e
-      ${CHGRP_CMD} *${rtype}*
-      ${STRICT_ON:-set -e}
+      for rfile in *"${rtype}"*; do 
+         if [[ -s "${rfile}" ]]; then
+            ${CHGRP_CMD} "${rfile}"
+         fi
+	 ${STRICT_ON:-set -e}
```
Comment (aerorahul) [inline scripts/exglobal_diag.sh]: Why is this needed?

Comment (WalterKolczynski-NOAA) [reply scripts/exglobal_diag.sh]: If `set +e` is removed, it's not. There is no need to turn back on something you never turned off.

Comment (HenryRWinterbottom) [reply scripts/exglobal_diag.sh]: @WalterKolczynski-NOAA noted that `set +e` was no longer needed. Is this? I would assume not but @aerorahul noted above the following https://github.com/NOAA-EMC/global-workflow/pull/2688#discussion_r1640603759.

Comment (HenryRWinterbottom) [reply scripts/exglobal_diag.sh]: Fixed.

**Review by aerorahul — APPROVED** [review]

**Comment by WalterKolczynski-NOAA** [comment]
The `resolves` keyword should be removed from the description since this only covers a small portion of the issue in question.

**Review by WalterKolczynski-NOAA — APPROVED** [review]
Conditionally approved.

**Comment by HenryRWinterbottom** [comment]
> The `resolves` keyword should be removed from the description since this only covers a small portion of the issue in question.

Changed to *References*.
